Tests/MarkCodableTests/Utility/TestModels.swift

```diff
@@ -117,3 +117,11 @@ struct Lists: Codable, Equatable {
 struct ListContainer<T: Codable>: Codable {
     let numbers: Array<T>
 }
+
+struct AnimalFarm: Codable, Equatable {
+    let pig: Pig
```

Comment:
@DivineDominion could we please use this opportunity to flex a few more code paths here? Looking at `MarkKeyedDecoding` I can think of 1) adding an optional `Pig` to walk the optional decoding paths 2) adding a optional property inside Pig set to `nil` to walk the decodeNil method.

```
    let pig: Pig
    let optionalPig: Pig?
```

and then

```
struct Pig ... {
  let name: String
  let color: String? 
}
```

Reply:
This is quite tough! (And now the scope of the PR changes accidentally.)

I've devised a test, but I'm not sure which variant is actually expected.

TLDR: @icanzilb do we want to add column headers for columns that only have `nil` values? Or do we want to skip them?

## Testing empty column headers are added ❌ 

This fails:

```swift
    func testSingleOptionalString() throws {
        struct SingleOptionalString: Codable, Equatable {
            var optionalString: String?
        }

        let encoder = MarkEncoder()
        let decoder = MarkDecoder()

        let singleNilMarkdown = """
        |optionalString|
        |--------------|
        |              |
        """
        let nilValue = SingleOptionalString(optionalString: nil)
        XCTAssertEqual(singleNilMarkdown, try encoder.encode(nilValue))
        XCTAssertEqual(nilValue, try decoder.decode(SingleOptionalString.self, string: singleNilMarkdown))

        let singleNonNilMarkdown = """
        |optionalString|
        |--------------|
        |yes           |
        """
        let existingValue = SingleOptionalString(optionalString: "yes")
        XCTAssertEqual(singleNonNilMarkdown, try encoder.encode(existingValue))
        XCTAssertEqual(existingValue, try decoder.decode(SingleOptionalString.self, string: singleNonNilMarkdown))
    }
```

## Same test, without header

Difference is:

```swift
        let singleNilMarkdown = """
        ||
        ||
        ||
        """
```

That does pass, but it's weird: why do we get a table at all, why no separator lines? I'm not sure if this is okay or not.

Reply:
I would add the columns that only have nil values **if we are able to** — this will give chance to humans to enter values in a plain text editor. I didn't want to expand the PR scope with my comment though, I thought it'd be trivial to expand the test 🙈

Do you feel like looking into the `nil` value issue in here? Otherwise we can merge and have a separate issue for this - up to you

Reply:
@icanzilb When do you think we would not (or should not?) be able to?

Reply:
@DivineDominion I'm not certain the encoder walks all the properties if they are set to `nil`.

Reply:
@icanzilb I got this working now, fixing #1 in the process (which is similar to what I've posted above)

Turns out we need `*IfPresent` for all primitive types. Yikes. Leaving that for a follow-up.